pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,188 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pleg
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	internalapi "k8s.io/cri-api/pkg/apis"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/utils/clock"
+)
+
+type EventedPLEG struct {
+	// The container runtime.
+	runtime kubecontainer.Runtime
+
+	runtimeService internalapi.RuntimeService
+
+	// The channel from which the subscriber listens events.
+	eventChannel chan *PodLifecycleEvent
+
+	// Cache for storing the runtime states required for syncing pods.
+	cache kubecontainer.Cache
+	// For testability.
+	clock clock.Clock
+}
+
+// NewGenericPLEG instantiates a new GenericPLEG object and return it.
+func NewEventedPLEG(runtime kubecontainer.Runtime, runtimeService internalapi.RuntimeService, eventChannel chan *PodLifecycleEvent,
+	cache kubecontainer.Cache, clock clock.Clock) PodLifecycleEventGenerator {
+	return &EventedPLEG{
+		runtime:        runtime,
+		runtimeService: runtimeService,
+		eventChannel:   eventChannel,
+		cache:          cache,
+		clock:          clock,
+	}
+}
+
+// Watch returns a channel from which the subscriber can receive PodLifecycleEvent
+// events.
+// TODO: support multiple subscribers.
+func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
+	return e.eventChannel
+}
+
+// Start spawns a goroutine to relist periodically.
+func (e *EventedPLEG) Start() {
+	go wait.Forever(e.watchEventsChannel, 0)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) >= cap(e.eventChannel) {
+		return false, fmt.Errorf("pleg event channel capacity is full with %v events", len(e.eventChannel))
+	}
+	return true, nil
+}
+
+func (e *EventedPLEG) watchEventsChannel() {
+
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
+	defer close(containerEventsResponseCh)
+
+	// Get the container events from the runtime.
+	go e.runtimeService.GetContainerEvents(containerEventsResponseCh)
+
+	e.processCRIEvents(containerEventsResponseCh)
+}
+
+func (e *EventedPLEG) processCRIEvents(containerEventsResponseCh chan *runtimeapi.ContainerEventResponse) {
+	for event := range containerEventsResponseCh {
+		switch event.ContainerEventType {
+		case runtimeapi.ContainerEventType_CONTAINER_STOPPED_EVENT:
+			e.updatePodStatus(event)
```

Comment:
nit: in every case you're calling `updatePodStatus`, we could consolidate above by callig before the switch statement